evidence.go

```diff
@@ -34,139 +34,163 @@ type JSONCollection struct {
+func DecodeAndValidateEvidenceFromCBOR(buf []byte) (*Evidence, error) {
+	ev, err := DecodeEvidenceFromCBOR(buf)
+	if err != nil {
+		return nil, err
+	}
 
-func (e *Evidence) MarshalJSON() ([]byte, error) {
-	if e.PlatformClaims == nil || e.RealmClaims == nil {
-		return nil, errors.New("invalid evidence")
+	if err := ev.Validate(); err != nil {
+		return nil, err
 	}
 
-	pj, err := e.PlatformClaims.ToJSON()
-	if err != nil {
-		return nil, fmt.Errorf("error serializing platform claims: %w", err)
+	return ev, nil
+}
+
+// DecodeEvidenceFromCBOR unmarshals CCA claims collection from provided CBOR.
+func DecodeEvidenceFromCBOR(buf []byte) (*Evidence, error) {
+	ev := &Evidence{}
+
+	if err := ev.UnmarshalCBOR(buf); err != nil {
+		return nil, err
 	}
 
-	rj, err := e.RealmClaims.ToJSON()
+	return ev, nil
+}
+
+// DecodeAndValidateEvidenceFromJSON unmarshals CCA claims collection from
+// provided JSON and validates both sets of claims.
+func DecodeAndValidateEvidenceFromJSON(buf []byte) (*Evidence, error) {
+	ev, err := DecodeEvidenceFromJSON(buf)
 	if err != nil {
-		return nil, fmt.Errorf("error serializing realm claims: %w", err)
+		return nil, err
 	}
 
-	c := JSONCollection{
-		PlatformToken: pj,
-		RealmToken:    rj,
+	if err := ev.Validate(); err != nil {
+		return nil, err
 	}
 
-	return json.Marshal(c)
+	return ev, nil
 }
 
-func (e *Evidence) MarshalUnvalidatedJSON() ([]byte, error) {
-	var pj, rj []byte
-	var err error
+// DecodeEvidenceFromJSON unmarshals CCA claims collection from provided JSON.
+func DecodeEvidenceFromJSON(buf []byte) (*Evidence, error) {
+	ev := &Evidence{}
 
-	if e.PlatformClaims != nil {
-		pj, err = e.PlatformClaims.ToUnvalidatedJSON()
-		if err != nil {
-			return nil, fmt.Errorf("error serializing platform claims: %w", err)
-		}
+	if err := ev.UnmarshalJSON(buf); err != nil {
+		return nil, err
 	}
 
-	if e.RealmClaims != nil {
-		rj, err = e.RealmClaims.ToUnvalidatedJSON()
-		if err != nil {
-			return nil, fmt.Errorf("error serializing realm claims: %w", err)
-		}
-	}
+	return ev, nil
+}
 
-	c := JSONCollection{
-		PlatformToken: pj,
-		RealmToken:    rj,
+// ValidateAndEncodeEvidenceToJSON validates and then marshals CCA platform claims
+// to JSON.
+func ValidateAndEncodeEvidenceToJSON(e *Evidence) ([]byte, error) {
+	if err := e.Validate(); err != nil {
+		return nil, err
 	}
 
-	return json.Marshal(c)
+	return EncodeEvidenceToJSON(e)
 }
 
-func (e *Evidence) UnmarshalJSON(data []byte) error {
-	p, r, err := e.doUnmarshalJSON(data)
-	if err != nil {
-		return err
-	}
+// EncodeEvidenceToJSON marshals CCA evidence to JSON.
+func EncodeEvidenceToJSON(e *Evidence) ([]byte, error) {
+	return json.Marshal(e)
+}
 
-	if p == nil {
-		return errors.New("unmarshaling CCA claims: missing platform claims")
+// Evidence is a wrapper around CcaToken
+type Evidence struct {
+	PlatformClaims platform.IClaims
+	RealmClaims    realm.IClaims
+	collection     *CBORCollection
+}
 
+func (e *Evidence) Validate() error {
```

Comment:
As Evidence is a composite structure, would help by adding a comment

// Validate checks both Platform Claims and Realm Claims for Validity